eth/executionclient/execution_client.go

```diff
@@ -234,6 +227,72 @@ func (ec *ExecutionClient) fetchLogsInBatches(ctx context.Context, startBlock, e
 	return logCh, errCh
 }
 
+// subdivideLogFetch handles log fetching with automatic subdivision on query limit errors.
+// It first attempts a direct fetch, and if that fails with a query limit error, it subdivides
+// the block range and tries again with smaller chunks.
+func (ec *ExecutionClient) subdivideLogFetch(ctx context.Context, q ethereum.FilterQuery) ([]ethtypes.Log, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case <-ec.closed:
+		return nil, ErrClosed
+	default:
+	}
+
+	logs, err := ec.client.FilterLogs(ctx, q)
+	if err == nil {
+		return logs, nil
+	}
+
+	if !isRPCQueryLimitError(err) || q.FromBlock == nil || q.ToBlock == nil {
+		return nil, err
+	}
+
+	fromBlock := q.FromBlock.Uint64()
+	toBlock := q.ToBlock.Uint64()
+
+	// require at least 2 blocks to subdivide (fromBlock must be less than toBlock)
+	if fromBlock >= toBlock {
+		return nil, fmt.Errorf("insufficient blocks to subdivide (fromBlock: %d, toBlock: %d): %w", fromBlock, toBlock, err)
+	}
+
+	ec.logger.Info("subdividing log fetch due to query limit",
+		fields.FromBlock(fromBlock),
+		fields.ToBlock(toBlock))
+
+	midBlock := fromBlock + (toBlock-fromBlock)/2
+
+	leftQuery := q
+	leftQuery.FromBlock = new(big.Int).SetUint64(fromBlock)
+	leftQuery.ToBlock = new(big.Int).SetUint64(midBlock)
+
+	rightQuery := q
+	rightQuery.FromBlock = new(big.Int).SetUint64(midBlock + 1)
+	rightQuery.ToBlock = new(big.Int).SetUint64(toBlock)
+
+	leftLogs, leftErr := ec.subdivideLogFetch(ctx, leftQuery)
+	if leftErr != nil {
+		return nil, leftErr
+	}
+
+	rightLogs, rightErr := ec.subdivideLogFetch(ctx, rightQuery)
+	if rightErr != nil {
+		return nil, rightErr
+	}
+
+	totalLogs := len(leftLogs) + len(rightLogs)
+	combinedLogs := make([]ethtypes.Log, 0, totalLogs)
```

Comment:
can we just `return append(leftLogs, rightLogs...)`?

Reply:
I see, I didn't think about it. Yes, I agree that pre-allocating would be good here. Can you please revert it? Sorry